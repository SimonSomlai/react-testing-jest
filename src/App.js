import React, {Component} from 'react';
import './App.css';
import Note from "./components/note.js"
import API from "./services/api.js"

class App extends Component {
  constructor(props) {
    super(props)
    this.state = {
      notes: [
        {
          id: 11,
          title: "First Tas",
          description: "Pick Up Paycheck"
        }, {
          id: 2,
          title: "Second Task",
          description: "Cash Paycheck"
        }, {
          id: 3,
          title: "Third Task",
          description: "Get Some Milk"
        }
      ]
    }
  }

  // Regular CRUD Stuff
  findNote(id) {
    return this.state.notes.find((item) => {
      if (item.id === id) {
        return item
      }
    })
  }

  findIndex(note){
    return this.state.notes.indexOf(note)
  }

  updateNote(newNote) {
    var note = this.findNote(newNote.id)
    var index = this.findIndex(note)
    this.state.notes[index] = newNote
    this.setState({notes: this.state.notes})
  }

  deleteNote(id) {
    var note = this.findNote(id)
    var index = this.state.notes.indexOf(note)
    this.state.notes.splice(index, 1)
    this.setState({notes: this.state.notes})
  }

  createNote(e) {
    e.preventDefault()
    var note = {}
    var fields = this.refs
    for (var prop in fields) {
      note[prop] = fields[prop].value
    }
    var id = this.state.notes.length + 1
    note.id = id
    this.state.notes.push(note)
    this.setState({notes: this.state.notes})
  }

  renderNotes() {
    var noteComponents = []
    this.state.notes.map((note, i) => {
      noteComponents.push(<Note deleteNote={(id) => this.deleteNote(id)} updateNote={(note) => this.updateNote(note)} key={i} note={note}/>)
    })
    return noteComponents
  }

  execute(){
    var apiPromise = API.getAllDinosaurs()
    apiPromise.then((data) => {console.log('data', data)})
  }

  render() {
    var styling = {
      container: {
        marginTop: "100px"
      },
      heading: {
        marginBottom: "50px"
      }
    }


    return (
      <div style={styling.container} className="container">

        <h1>Create a new note</h1>
        <form>
          <input className="note-title form-control form-inline" ref="title" type="text"/>
          <textarea className="note-description form-control form-inline" ref="description"></textarea>
          <button type="submit" className="create-note" onClick={(e) => this.createNote(e)}>Create Note</button>
        </form>
        <div className="row">
          <h1 style={styling.heading} className="text-center">My Notes</h1>
          {this.renderNotes()}
        </div>
      </div>
    );
  }
}

export default App;
