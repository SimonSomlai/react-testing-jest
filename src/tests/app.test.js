
import App from "../App.js";
import Note from "../components/note.js";

import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import toJson from "enzyme-to-json"
import {mount} from "enzyme";

// GENERAL TESTING PATTERN
// - setup > action > verify action got desired result
// Why should you test? -> When refactoring/extending a big application, you might break certain features. When
// you've tested your code, you'll know which features break + the location they break at in the code.

// What should be tested? -> Depends on the type of test. Unit tests test one function which takes an input and verify they got
// valid output (most useful for complex algorithms). Integration tests test a feature/user story and ensure it works.

// CHEAT SHEET
// component.instance().props
// component.state()
// component.instance()  .updateNote(note); = this
// component.find('some-css')
// button.simulate("click")
// component.find('.note-title').node.value = "new title";

// jest.spyOn(Note.prototype, 'toggleState');
// expect(toggleState).toHaveBeenCalled()

describe('ToDo App', () => {
  let note,
    component,
    notes
  beforeEach(() => {
    // Set starting global variables
    note = {
      id: 2,
      title: 'test note',
      description: "blabla"
    };
    notes = [
      {
        id: 1,
        title: "First Task",
        description: "Pick Up Paycheck"
      }, {
        id: 2,
        title: "Second Task",
        description: "Cash Paycheck"
      }, {
        id: 3,
        title: "Third Task",
        description: "Get Some Milk"
      }
    ];
  });

  describe("INTEGRATION TESTS", () => {

    it("should render correctly", () => {
      const component = mount(<App/>);
      const tree = toJson(component)
      expect(tree).toMatchSnapshot()
    })

    it("UI should change to edit mode when edit mode is changed", () => {
      component = mount(<Note note={note}/>);
      component.setState({isEditing: true});

      //verify: a form has replaced the title and description with default value
      expect(component.find('.panel-title input')).toBeTruthy();
      expect(component.find('.panel-title input').node.value).toBe(note.title)
      expect(component.find('.panel-body textarea').node.value).toBe(note.description);
    })

    it("Should trigger toggleState when edit button is clicked & change state accordingly", () => {
      const toggleState = jest.spyOn(Note.prototype, 'toggleState');
      component = mount(<Note note={note}/>);
      component.find('.edit-button').simulate("click")
      expect(toggleState).toHaveBeenCalled()
    })

    it('Should create correctly in the state & show correctly on UI', () => {
      const component = mount(<App/>);
      var initialUiLength = component.find('.panel-title').length;
      var initialStateLength = component.state().notes.length;
      component.find('.note-title').node.value = "new title";
      component.find('.note-description').node.value = "new description";
      component.find('.create-note').simulate('click');

      const newNote = component.state().notes[component.state().notes.length - 1];
      expect(newNote.title).toBe("new title");
      expect(newNote.description).toBe("new description");

      expect(component.find('.panel-title').length).toBe(initialUiLength + 1); // Check if UI gets updated correctly
      expect(component.state().notes.length).toBe(initialStateLength + 1); // Check if state gets updated correctly
    })

    it('Should update correctly in the state & show correctly on UI', () => {
      const component = mount(<App/>);
      component.find('.edit-button').first().simulate('click'); // Simulate a click event
      expect(component.find('.save-button').length).toBe(1) // Save buttons should have appeared
      component.find('.panel-title input').node.value = "Changing the note!" // Simulate edit
      component.find('.save-button').simulate('click'); // Simulate a click event on the save button

      expect(component.find('.panel-title').first().text()).toBe("Changing the note!") // Check if text is set right on UI
      expect(component.state().notes[0].title).toBe("Changing the note!") // Check if text is set right in state
      expect(component.find('.save-button').length).toBe(0) // Save buttons should be gone
    })

    it('Should delete correctly in the state & show correctly on UI', () => {
      const component = mount(<App/>);
      var initialUiLength = component.find('.panel-title').length;
      var initialStateLength = component.state().notes.length;
      component.find('.delete-button').first().simulate('click'); // Simulate a click event
      expect(component.find('.panel-title').length).toBe(initialUiLength - 1); // Check if UI gets updated correctly
      expect(component.state().notes.length).toBe(initialStateLength - 1); // Check if state gets updated correctly
    })
  })

  describe("UNIT TESTS", () => {
    it('function findNote(id) should find the right note when given an id', () => {
      const component = mount(<App/>);
      component.setState({notes: notes});
      expect(component.instance().findNote(1)).toBe(notes[0]);
    })

    it('function findIndex(note) should retrieve the right index in the state', () => {
      const component = mount(<App/>);
      component.setState({notes: notes});
      expect(component.instance().findIndex(component.state().notes[2])).toBe(2);
    })

    it('function updateNote(newNote) should update the note correctly', () => {
      const component = mount(<App/>);
      component.setState({notes: notes});
      component.instance().updateNote(note);
      var index = component.instance().findIndex(note);
      expect(component.state().notes[index]).toBe(note);
    })
  })
});
