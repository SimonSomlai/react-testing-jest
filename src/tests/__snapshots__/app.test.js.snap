// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo App INTEGRATION TESTS should render 1`] = `
<App>
  <div
    className="container"
    style={
      Object {
        "marginTop": "100px",
      }
    }
  >
    <h1>
      Create a new note
    </h1>
    <form>
      <input
        className="note-title form-control form-inline"
        type="text"
      />
      <textarea
        className="note-description form-control form-inline"
      />
      <button
        className="create-note"
        onClick={[Function]}
        type="submit"
      >
        Create Note
      </button>
    </form>
    <div
      className="row"
    >
      <h1
        className="text-center"
        style={
          Object {
            "marginBottom": "50px",
          }
        }
      >
        My Notes
      </h1>
      <Note
        deleteNote={[Function]}
        note={
          Object {
            "description": "Pick Up Paycheck",
            "id": 1,
            "title": "First Task",
          }
        }
        updateNote={[Function]}
      >
        <div
          className="col-md-4"
        >
          <div
            className="panel panel-default"
          >
            <div
              className="panel-heading"
            >
              <h3
                className="panel-title"
              >
                First Task
              </h3>
            </div>
            <div
              className="panel-body"
            >
              Pick Up Paycheck
            </div>
            <button
              className="btn btn-primary edit-button"
              onClick={[Function]}
            >
              Edit Note
            </button>
            <button
              className="btn btn-primary delete-button"
              onClick={[Function]}
            >
              Delete Note
            </button>
          </div>
        </div>
      </Note>
      <Note
        deleteNote={[Function]}
        note={
          Object {
            "description": "Cash Paycheck",
            "id": 2,
            "title": "Second Task",
          }
        }
        updateNote={[Function]}
      >
        <div
          className="col-md-4"
        >
          <div
            className="panel panel-default"
          >
            <div
              className="panel-heading"
            >
              <h3
                className="panel-title"
              >
                Second Task
              </h3>
            </div>
            <div
              className="panel-body"
            >
              Cash Paycheck
            </div>
            <button
              className="btn btn-primary edit-button"
              onClick={[Function]}
            >
              Edit Note
            </button>
            <button
              className="btn btn-primary delete-button"
              onClick={[Function]}
            >
              Delete Note
            </button>
          </div>
        </div>
      </Note>
      <Note
        deleteNote={[Function]}
        note={
          Object {
            "description": "Get Some Milk",
            "id": 3,
            "title": "Third Task",
          }
        }
        updateNote={[Function]}
      >
        <div
          className="col-md-4"
        >
          <div
            className="panel panel-default"
          >
            <div
              className="panel-heading"
            >
              <h3
                className="panel-title"
              >
                Third Task
              </h3>
            </div>
            <div
              className="panel-body"
            >
              Get Some Milk
            </div>
            <button
              className="btn btn-primary edit-button"
              onClick={[Function]}
            >
              Edit Note
            </button>
            <button
              className="btn btn-primary delete-button"
              onClick={[Function]}
            >
              Delete Note
            </button>
          </div>
        </div>
      </Note>
    </div>
  </div>
</App>
`;

exports[`ToDo App INTEGRATION TESTS should render correctly 1`] = `
<App>
  <div
    className="container"
    style={
      Object {
        "marginTop": "100px",
      }
    }
  >
    <h1>
      Create a new note
    </h1>
    <form>
      <input
        className="note-title form-control form-inline"
        type="text"
      />
      <textarea
        className="note-description form-control form-inline"
      />
      <button
        className="create-note"
        onClick={[Function]}
        type="submit"
      >
        Create Note
      </button>
    </form>
    <div
      className="row"
    >
      <h1
        className="text-center"
        style={
          Object {
            "marginBottom": "50px",
          }
        }
      >
        My Notes
      </h1>
      <Note
        deleteNote={[Function]}
        note={
          Object {
            "description": "Pick Up Paycheck",
            "id": 1,
            "title": "First Task",
          }
        }
        updateNote={[Function]}
      >
        <div
          className="col-md-4"
        >
          <div
            className="panel panel-default"
          >
            <div
              className="panel-heading"
            >
              <h3
                className="panel-title"
              >
                First Task
              </h3>
            </div>
            <div
              className="panel-body"
            >
              Pick Up Paycheck
            </div>
            <button
              className="btn btn-primary edit-button"
              onClick={[Function]}
            >
              Edit Note
            </button>
            <button
              className="btn btn-primary delete-button"
              onClick={[Function]}
            >
              Delete Note
            </button>
          </div>
        </div>
      </Note>
      <Note
        deleteNote={[Function]}
        note={
          Object {
            "description": "Cash Paycheck",
            "id": 2,
            "title": "Second Task",
          }
        }
        updateNote={[Function]}
      >
        <div
          className="col-md-4"
        >
          <div
            className="panel panel-default"
          >
            <div
              className="panel-heading"
            >
              <h3
                className="panel-title"
              >
                Second Task
              </h3>
            </div>
            <div
              className="panel-body"
            >
              Cash Paycheck
            </div>
            <button
              className="btn btn-primary edit-button"
              onClick={[Function]}
            >
              Edit Note
            </button>
            <button
              className="btn btn-primary delete-button"
              onClick={[Function]}
            >
              Delete Note
            </button>
          </div>
        </div>
      </Note>
      <Note
        deleteNote={[Function]}
        note={
          Object {
            "description": "Get Some Milk",
            "id": 3,
            "title": "Third Task",
          }
        }
        updateNote={[Function]}
      >
        <div
          className="col-md-4"
        >
          <div
            className="panel panel-default"
          >
            <div
              className="panel-heading"
            >
              <h3
                className="panel-title"
              >
                Third Task
              </h3>
            </div>
            <div
              className="panel-body"
            >
              Get Some Milk
            </div>
            <button
              className="btn btn-primary edit-button"
              onClick={[Function]}
            >
              Edit Note
            </button>
            <button
              className="btn btn-primary delete-button"
              onClick={[Function]}
            >
              Delete Note
            </button>
          </div>
        </div>
      </Note>
    </div>
  </div>
</App>
`;
